mp/src/game/server/neo/neo_player.cpp

```diff
@@ -1768,8 +1766,54 @@ void CNEO_Player::Event_Killed( const CTakeDamageInfo &info )
 	}
 
 	BaseClass::Event_Killed(info);
+}
+
+void CNEO_Player::Weapon_DropOnDeath(CBaseCombatWeapon* pWeapon, Vector velocity)
+{
+	CNEOBaseCombatWeapon* pNeoWeapon = static_cast<CNEOBaseCombatWeapon*>(pWeapon);
+	if (!pNeoWeapon)
+		return;
+	if (!pNeoWeapon->GetWpnData().m_bDropOnDeath)
+		return;
+
+	// If attack button was held down when player died, drop a live grenade. NEOTODO (Adam) Add delay between pressing an attack button and the pin being fully pulled out. If pin not out when dead do not drop a live grenade
+	if (pNeoWeapon->GetNeoWepBits() & NEO_WEP_FRAG_GRENADE) {
+		if (((m_nButtons & IN_ATTACK) && (!(m_afButtonPressed & IN_ATTACK))) ||
+			((m_nButtons & IN_ATTACK2) && (!(m_afButtonPressed & IN_ATTACK2)))) {
+			auto pWeaponFrag = static_cast<CWeaponGrenade*>(pNeoWeapon);
+			pWeaponFrag->DropLiveGrenade(this, velocity);
+			return;
+		}
+	}
+	if (pNeoWeapon->GetNeoWepBits() & NEO_WEP_SMOKE_GRENADE) {
+		if (((m_nButtons & IN_ATTACK) && (!(m_afButtonPressed & IN_ATTACK))) ||
+			((m_nButtons & IN_ATTACK2) && (!(m_afButtonPressed & IN_ATTACK2)))) {
+			auto pWeaponSmoke = static_cast<CWeaponSmokeGrenade*>(pNeoWeapon);
+			pWeaponSmoke->DropLiveGrenade(this, velocity);
+			return;
+		}
+	}
```

Comment:
I think the code around `DropLiveGrenade` can be improved.

- `CheckThrowPosition` used within both grenade's `DropLiveGrenade` doesn't actually utilize any of the method var/func within and is repeated the same twice.  It can be separated out as it own plain function.
- `DropLiveGrenade` body could then be pulled out into this scope where it can be fitted in with the common codes and only differ a bit when creating grenade, and the two `DropLiveGrenade` can be dropped from being a function method.
- The if statement can then be combind with: `... & NEO_WEP_FRAG_GRENADE | NEO_WEP_SMOKE_GRENADE`

Reply:
DropLiveGrenade is mostly a copy of the existing ThrowGrenade function except the initial velocity of the grenade is the player's velocity at the time of death, CheckThrowPosition potentially modifies the value of vecSrc which is then used in NEOFraggrenade_Create/NEOSmokegrenade_Create

![image](https://github.com/user-attachments/assets/b9ab421e-23d0-47ee-8ae3-72e1ea06562c)
I like the idea of including the content of DropLiveGrenade once in here, using for example isFrag to change the behaviour slightly between the two but I would have to drop CheckThrowPosition, and since I no longer see the sv_neo_infinite_frag/smoke_grenades I would first have to call a function to find that server convar to then check its value, unless we just skip that check since its probably not necessary anyway, you loose the weapon_grenade/smokegrenade the moment you throw a grenade

Reply:
I could just modify ThrowGrenade to take an optional parameter deathDrop or something, or even just checking if owner is alive when throwing to not include the players throw force in vecThrow

Reply:
actually I just remembered there is a check for if player is alive, but when weapon drop is called the player is still technically alive, hence the new function